syntax = "proto3";

//
// planner request
//

message DesiredMagazineState {
  // which sequence of products is expected in this magazine (at most 3)
  // a product is either red, white, or blue
  repeated string output = 1;
}

message PlannerRequest {
  // path to ontology that provides the topology
  string ontology = 2;
  // which magazine states are expected (at most 2)
  repeated DesiredMagazineState goal = 1;
  // maximum number of time steps to consider (at most 20)
  uint32 max_time_step = 3;
}

message SkillMatchTriple {
  // classes in the ontology
  // = what we require from the machine
  // e.g. [UnsealedBox,BlueSeal] (as inputs)
  repeated string requirement = 1;

  // a class in the ontology
  // = what the machine offers
  // e.g. [BoxClosewithBlueSealSkill,BlueSealedBox] (as skill + outputs)
  repeated string offer = 2;

  // an instance in the ontology
  // = the machine that offers or does not provide the offer under the requirement
  // e.g. "sealerunit1"
  string module = 3;
}

message SkillMatchResult {
  // all pairs of requirements/offers that match
  repeated SkillMatchTriple positiveMatches = 1;
  // all pairs of requirements/offers that do not
  repeated SkillMatchTriple negativeMatches = 2;
}

message SkillMatchResultAndPlannerRequest {
  SkillMatchResult skillmatchResult = 1;
  PlannerRequest plannerRequest = 2;
}

//
// pipeline result
//

message Action {
  uint32 timestep = 1;
  string action = 2;
}

message PlannerResult {
  repeated Action actions = 1;
}

message ProductInput {
  string product = 1; // name of the product, for identification in result
  string type = 2;    // red, white, blue
  float timeEnteringSystem = 3; // seconds
}

message TimeEstimateRequest {
  repeated ProductInput products_entering_system = 1;
}

message PlannerResultAndTimeEstimateRequest {
  PlannerResult plannerResult = 1;
  TimeEstimateRequest timeEstimateRequest = 2;
}

service AI4IndustryPlanner {
  rpc computeOptimalPlan(SkillMatchResultAndPlannerRequest) returns (PlannerResultAndTimeEstimateRequest);
}
